{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6ac3cea5",
   "metadata": {},
   "source": [
    "## Heyyyo lets put our algo's here!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0440684f",
   "metadata": {},
   "source": [
    "# Blosum62"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "857712f3",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "blosum62 = {\n",
    "    ('W', 'F'): 1, ('L', 'R'): -2, ('S', 'P'): -1, ('V', 'T'): 0,\n",
    "    ('Q', 'Q'): 5, ('N', 'A'): -2, ('Z', 'Y'): -2, ('W', 'R'): -3,\n",
    "    ('Q', 'A'): -1, ('S', 'D'): 0, ('H', 'H'): 8, ('S', 'H'): -1,\n",
    "    ('H', 'D'): -1, ('L', 'N'): -3, ('W', 'A'): -3, ('Y', 'M'): -1,\n",
    "    ('G', 'R'): -2, ('Y', 'I'): -1, ('Y', 'E'): -2, ('B', 'Y'): -3,\n",
    "    ('Y', 'A'): -2, ('V', 'D'): -3, ('B', 'S'): 0, ('Y', 'Y'): 7,\n",
    "    ('G', 'N'): 0, ('E', 'C'): -4, ('Y', 'Q'): -1, ('Z', 'Z'): 4,\n",
    "    ('V', 'A'): 0, ('C', 'C'): 9, ('M', 'R'): -1, ('V', 'E'): -2,\n",
    "    ('T', 'N'): 0, ('P', 'P'): 7, ('V', 'I'): 3, ('V', 'S'): -2,\n",
    "    ('Z', 'P'): -1, ('V', 'M'): 1, ('T', 'F'): -2, ('V', 'Q'): -2,\n",
    "    ('K', 'K'): 5, ('P', 'D'): -1, ('I', 'H'): -3, ('I', 'D'): -3,\n",
    "    ('T', 'R'): -1, ('P', 'L'): -3, ('K', 'G'): -2, ('M', 'N'): -2,\n",
    "    ('P', 'H'): -2, ('F', 'Q'): -3, ('Z', 'G'): -2, ('X', 'L'): -1,\n",
    "    ('T', 'M'): -1, ('Z', 'C'): -3, ('X', 'H'): -1, ('D', 'R'): -2,\n",
    "    ('B', 'W'): -4, ('X', 'D'): -1, ('Z', 'K'): 1, ('F', 'A'): -2,\n",
    "    ('Z', 'W'): -3, ('F', 'E'): -3, ('D', 'N'): 1, ('B', 'K'): 0,\n",
    "    ('X', 'X'): -1, ('F', 'I'): 0, ('B', 'G'): -1, ('X', 'T'): 0,\n",
    "    ('F', 'M'): 0, ('B', 'C'): -3, ('Z', 'I'): -3, ('Z', 'V'): -2,\n",
    "    ('S', 'S'): 4, ('L', 'Q'): -2, ('W', 'E'): -3, ('Q', 'R'): 1,\n",
    "    ('N', 'N'): 6, ('W', 'M'): -1, ('Q', 'C'): -3, ('W', 'I'): -3,\n",
    "    ('S', 'C'): -1, ('L', 'A'): -1, ('S', 'G'): 0, ('L', 'E'): -3,\n",
    "    ('W', 'Q'): -2, ('H', 'G'): -2, ('S', 'K'): 0, ('Q', 'N'): 0,\n",
    "    ('N', 'R'): 0, ('H', 'C'): -3, ('Y', 'N'): -2, ('G', 'Q'): -2,\n",
    "    ('Y', 'F'): 3, ('C', 'A'): 0, ('V', 'L'): 1, ('G', 'E'): -2,\n",
    "    ('G', 'A'): 0, ('K', 'R'): 2, ('E', 'D'): 2, ('Y', 'R'): -2,\n",
    "    ('M', 'Q'): 0, ('T', 'I'): -1, ('C', 'D'): -3, ('V', 'F'): -1,\n",
    "    ('T', 'A'): 0, ('T', 'P'): -1, ('B', 'P'): -2, ('T', 'E'): -1,\n",
    "    ('V', 'N'): -3, ('P', 'G'): -2, ('M', 'A'): -1, ('K', 'H'): -1,\n",
    "    ('V', 'R'): -3, ('P', 'C'): -3, ('M', 'E'): -2, ('K', 'L'): -2,\n",
    "    ('V', 'V'): 4, ('M', 'I'): 1, ('T', 'Q'): -1, ('I', 'G'): -4,\n",
    "    ('P', 'K'): -1, ('M', 'M'): 5, ('K', 'D'): -1, ('I', 'C'): -1,\n",
    "    ('Z', 'D'): 1, ('F', 'R'): -3, ('X', 'K'): -1, ('Q', 'D'): 0,\n",
    "    ('X', 'G'): -1, ('Z', 'L'): -3, ('X', 'C'): -2, ('Z', 'H'): 0,\n",
    "    ('B', 'L'): -4, ('B', 'H'): 0, ('F', 'F'): 6, ('X', 'W'): -2,\n",
    "    ('B', 'D'): 4, ('D', 'A'): -2, ('S', 'L'): -2, ('X', 'S'): 0,\n",
    "    ('F', 'N'): -3, ('S', 'R'): -1, ('W', 'D'): -4, ('V', 'Y'): -1,\n",
    "    ('W', 'L'): -2, ('H', 'R'): 0, ('W', 'H'): -2, ('H', 'N'): 1,\n",
    "    ('W', 'T'): -2, ('T', 'T'): 5, ('S', 'F'): -2, ('W', 'P'): -4,\n",
    "    ('L', 'D'): -4, ('B', 'I'): -3, ('L', 'H'): -3, ('S', 'N'): 1,\n",
    "    ('B', 'T'): -1, ('L', 'L'): 4, ('Y', 'K'): -2, ('E', 'Q'): 2,\n",
    "    ('Y', 'G'): -3, ('Z', 'S'): 0, ('Y', 'C'): -2, ('G', 'D'): -1,\n",
    "    ('B', 'V'): -3, ('E', 'A'): -1, ('Y', 'W'): 2, ('E', 'E'): 5,\n",
    "    ('Y', 'S'): -2, ('C', 'N'): -3, ('V', 'C'): -1, ('T', 'H'): -2,\n",
    "    ('P', 'R'): -2, ('V', 'G'): -3, ('T', 'L'): -1, ('V', 'K'): -2,\n",
    "    ('K', 'Q'): 1, ('R', 'A'): -1, ('I', 'R'): -3, ('T', 'D'): -1,\n",
    "    ('P', 'F'): -4, ('I', 'N'): -3, ('K', 'I'): -3, ('M', 'D'): -3,\n",
    "    ('V', 'W'): -3, ('W', 'W'): 11, ('M', 'H'): -2, ('P', 'N'): -2,\n",
    "    ('K', 'A'): -1, ('M', 'L'): 2, ('K', 'E'): 1, ('Z', 'E'): 4,\n",
    "    ('X', 'N'): -1, ('Z', 'A'): -1, ('Z', 'M'): -1, ('X', 'F'): -1,\n",
    "    ('K', 'C'): -3, ('B', 'Q'): 0, ('X', 'B'): -1, ('B', 'M'): -3,\n",
    "    ('F', 'C'): -2, ('Z', 'Q'): 3, ('X', 'Z'): -1, ('F', 'G'): -3,\n",
    "    ('B', 'E'): 1, ('X', 'V'): -1, ('F', 'K'): -3, ('B', 'A'): -2,\n",
    "    ('X', 'R'): -1, ('D', 'D'): 6, ('W', 'G'): -2, ('Z', 'F'): -3,\n",
    "    ('S', 'Q'): 0, ('W', 'C'): -2, ('W', 'K'): -3, ('H', 'Q'): 0,\n",
    "    ('L', 'C'): -1, ('W', 'N'): -4, ('S', 'A'): 1, ('L', 'G'): -4,\n",
    "    ('W', 'S'): -3, ('S', 'E'): 0, ('H', 'E'): 0, ('S', 'I'): -2,\n",
    "    ('H', 'A'): -2, ('S', 'M'): -1, ('Y', 'L'): -1, ('Y', 'H'): 2,\n",
    "    ('Y', 'D'): -3, ('E', 'R'): 0, ('X', 'P'): -2, ('G', 'G'): 6,\n",
    "    ('G', 'C'): -3, ('E', 'N'): 0, ('Y', 'T'): -2, ('Y', 'P'): -3,\n",
    "    ('T', 'K'): -1, ('A', 'A'): 4, ('P', 'Q'): -1, ('T', 'C'): -1,\n",
    "    ('V', 'H'): -3, ('T', 'G'): -2, ('I', 'Q'): -3, ('Z', 'T'): -1,\n",
    "    ('C', 'R'): -3, ('V', 'P'): -2, ('P', 'E'): -1, ('M', 'C'): -1,\n",
    "    ('K', 'N'): 0, ('I', 'I'): 4, ('P', 'A'): -1, ('M', 'G'): -3,\n",
    "    ('T', 'S'): 1, ('I', 'E'): -3, ('P', 'M'): -2, ('M', 'K'): -1,\n",
    "    ('I', 'A'): -1, ('P', 'I'): -3, ('R', 'R'): 5, ('X', 'M'): -1,\n",
    "    ('L', 'I'): 2, ('X', 'I'): -1, ('Z', 'B'): 1, ('X', 'E'): -1,\n",
    "    ('Z', 'N'): 0, ('X', 'A'): 0, ('B', 'R'): -1, ('B', 'N'): 3,\n",
    "    ('F', 'D'): -3, ('X', 'Y'): -1, ('Z', 'R'): 0, ('F', 'H'): -1,\n",
    "    ('B', 'F'): -3, ('F', 'L'): 0, ('X', 'Q'): -1, ('B', 'B'): 4\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ceca820",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "GAP = -5    # gap penalty, we can chnge this\n",
    "\n",
    "def score(a, b):\n",
    "    \"\"\"Return BLOSUM62 score for characters a and b.\"\"\"\n",
    "    if (a, b) in blosum62:\n",
    "        return blosum62[(a, b)]\n",
    "    if (b, a) in blosum62:  # matrix is symmetric\n",
    "        return blosum62[(b, a)]\n",
    "    raise KeyError(f\"No score for pair({a}, {b})\")\n",
    "\n",
    "def blosum_align(seq1, seq2):\n",
    "    \"\"\"Needleman-Wuncsch global alignment using BLOSUM62.\"\"\"\n",
    "    n = len(seq1)\n",
    "    m = len(seq2)\n",
    "\n",
    "    # DP matrices\n",
    "    dp = [[0]*(m+1) for _ in range(n+1)]\n",
    "    ptr = [[None]*(m+1) for _ in range(n+1)]    # traceback\n",
    "\n",
    "    # initialize boundaries\n",
    "    for i in range(1, n+1):\n",
    "        dp[i][0] = i * GAP\n",
    "        ptr[i][0] = 'U'\n",
    "    for j in range(1, m+1):\n",
    "        dp[0][j] = j * GAP\n",
    "        ptr[0][j] = 'L'\n",
    "\n",
    "    # fill DP\n",
    "    for i in range(1, n+1):\n",
    "        for j in range(1, m+1):\n",
    "            match = dp[i-1][j-1] + score(seq1[i-1], seq2[j-1])\n",
    "            delete = dp[i-1][j] + GAP\n",
    "            insert = dp[i][j-1] + GAP\n",
    "\n",
    "            dp[i][j] = max(match, delete, insert)\n",
    "\n",
    "            if dp[i][j] == match:\n",
    "                ptr[i][j] = 'D'\n",
    "            elif dp[i][j] == delete:\n",
    "                ptr[i][j] = 'U'\n",
    "            else:\n",
    "                ptr[i][j] = 'L'\n",
    "\n",
    "    # traceback\n",
    "    aligned1 = []\n",
    "    aligned2 = []\n",
    "    i, j = n, m\n",
    "\n",
    "    while i > 0 or j > 0:\n",
    "        move = ptr[i][j]\n",
    "\n",
    "        if move == 'D':\n",
    "            aligned1.append(seq1[i-1])\n",
    "            aligned2.append(seq2[j-1])\n",
    "            i-=1\n",
    "            j-=1\n",
    "        elif move == 'U':\n",
    "            aligned1.append(seq1[i-1])\n",
    "            aligned2.append('-')\n",
    "            i-=1\n",
    "        elif move == 'L':\n",
    "            aligned1.append('-')\n",
    "            aligned2.append(seq2[j-1])\n",
    "            j-=1\n",
    "    \n",
    "    return dp[n][m], ''.join(reversed(aligned1)), \"\".join(reversed(aligned2))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d1034eb",
   "metadata": {},
   "source": [
    "# Local vs Global (Smith Waterman vs Needleman-Wunsch)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75072e93",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "827e784e",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "def read_fasta(filename):\n",
    "    with open(filename, 'r') as f:\n",
    "        sequence = \"\"\n",
    "        for line in f:\n",
    "            line = line.strip()\n",
    "            if line.startswith(\">\"): #ignore first line with label\n",
    "                pass\n",
    "            else:\n",
    "                sequence += line\n",
    "    return sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb354ff3",
   "metadata": {},
   "source": [
    "## Smith Waterman"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f202d5fc",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "class SmithWaterman:\n",
    "    \"\"\"\n",
    "    Performs the Smith-Waterman local alignment algorithm.\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, seq1, seq2, match_score=1, mismatch_penalty=-1, gap_penalty=2):\n",
    "\n",
    "        self.seq1 = seq1\n",
    "        self.seq2 = seq2\n",
    "        self.match_score = match_score\n",
    "        self.mismatch_penalty = mismatch_penalty\n",
    "        self.gap_penalty = gap_penalty\n",
    "\n",
    "        self.rows = len(seq1) + 1\n",
    "        self.cols = len(seq2) + 1\n",
    "\n",
    "        self.max_score = 0\n",
    "        self.max_pos = (0, 0)\n",
    "\n",
    "        self.fill_matrix()\n",
    "\n",
    "        self.score = self.find_score()\n",
    "        self.alignment = self.traceback()\n",
    "    \n",
    "    def fill_matrix(self):\n",
    "\n",
    "        # Initialize the scoring matrix with zeros\n",
    "        self.matrix = np.zeros((self.rows, self.cols), dtype = int)\n",
    "\n",
    "        # Fill the scoring matrix\n",
    "        for i in range(1, self.rows):\n",
    "            for j in range(1, self.cols):\n",
    "                score_diag = self.matrix[i - 1, j - 1] + (self.match_score if self.seq1[i - 1] == self.seq2[j - 1] else self.mismatch_penalty)\n",
    "                score_up = self.matrix[i - 1, j] - self.gap_penalty\n",
    "                score_left = self.matrix[i, j - 1] - self.gap_penalty\n",
    "            \n",
    "                self.matrix[i, j] = max(0, score_diag, score_up, score_left)\n",
    "\n",
    "    def find_score(self):\n",
    "        for i in range(1, self.rows):\n",
    "            for j in range(1, self.cols):\n",
    "                if self.matrix[i, j] > self.max_score:\n",
    "                    self.max_score = self.matrix[i, j]\n",
    "                    self.max_pos = (i, j)\n",
    "\n",
    "        return self.max_score\n",
    "\n",
    "    def traceback(self):\n",
    "        i, j = self.max_pos\n",
    "        aligned_seq1, aligned_seq2 = \"\", \"\"\n",
    "        traceback_coords = []\n",
    "    \n",
    "        #Traceback and create the alignment\n",
    "        while self.matrix[i][j] != 0:\n",
    "            traceback_coords.append((i,j))\n",
    "            current_score = self.matrix[i][j]\n",
    "            if current_score == self.matrix[i-1][j-1] + (self.match_score if self.seq1[i-1] == self.seq2[j-1] else self.mismatch_penalty):\n",
    "                aligned_seq1 = self.seq1[i-1] + aligned_seq1\n",
    "                aligned_seq2 = self.seq2[j-1] + aligned_seq2\n",
    "                i -= 1\n",
    "                j -= 1\n",
    "            elif current_score == self.matrix[i-1][j] + self.gap_penalty:\n",
    "                aligned_seq1 = self.seq1[i-1] + aligned_seq1\n",
    "                aligned_seq2 = \"-\" + aligned_seq2\n",
    "                i -= 1\n",
    "            else:\n",
    "                aligned_seq1 = \"-\" + aligned_seq1\n",
    "                aligned_seq2 = self.seq2[j-1] + aligned_seq2\n",
    "                j -= 1\n",
    "                \n",
    "        self.traceback_coords = traceback_coords\n",
    "\n",
    "        while i > 0:\n",
    "            aligned_seq1 = self.seq1[i - 1] + aligned_seq1\n",
    "            aligned_seq2 = \"-\" + aligned_seq2\n",
    "            i -= 1\n",
    "\n",
    "        while j > 0:\n",
    "            aligned_seq1 = \"-\" + aligned_seq1\n",
    "            aligned_seq2 = self.seq2[j - 1] + aligned_seq2\n",
    "            j -= 1\n",
    "        \n",
    "        return aligned_seq1, aligned_seq2\n",
    "    \n",
    "    def display_matrix(self):\n",
    "        '''\n",
    "        Returns the scoring matrix with the traceback labeled in cyan\n",
    "        '''\n",
    "        df = pd.DataFrame(self.matrix) \n",
    "\n",
    "        styled_df = df.style.apply( lambda row: [\n",
    "            \"background-color: cyan\" \n",
    "            if (row.name, col_idx) in self.traceback_coords else \"\"\n",
    "            for col_idx in range(df.shape[1]) ], axis=1 )\n",
    "        \n",
    "\n",
    "        return styled_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99e2469e",
   "metadata": {},
   "source": [
    "## Needleman-Wunsch Algorithm\n",
    "\n",
    "This algorithm, developed by Saul B. Needleman and Christian D. Wunsch, is utilized for global sequence alignment of amino acids and nucleotides."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "237581c3",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "class NeedlemanWunsch:\n",
    "    '''\n",
    "    Contains the Needleman-Wunsch score & matrix for two sequences.\n",
    "\n",
    "    Important Attributes:\n",
    "    score (int) : the score for the matching algorithm\n",
    "    alignment (tuple (str, str)) : the alignments for sequence 1 and 2\n",
    "\n",
    "    Important Methods:\n",
    "    display_matrix() : displays the scored matrix as a DataFrame\n",
    "                       where light blue is the traceback method.\n",
    "    '''\n",
    "    def __init__(self, seq1, seq2, match = 1, mismatch = -1, gap = -1):\n",
    "        self.seq1 = seq1\n",
    "        self.seq2 = seq2\n",
    "        self.match = match\n",
    "        self.mismatch = mismatch\n",
    "        self.gap = gap\n",
    "    \n",
    "        self.rows = len(seq1) + 1\n",
    "        self.columns = len(seq2) + 1\n",
    "\n",
    "        self.fill_matrix()\n",
    "\n",
    "        self.score = self.find_score()\n",
    "        self.alignment = self.traceback()\n",
    "\n",
    "    def init_matrix(self):\n",
    "        self.matrix = np.zeros((self.rows, self.columns), dtype = int)\n",
    "        self.matrix[0, :] = np.arange(0, (self.columns) * self.gap, self.gap)\n",
    "        self.matrix[:, 0] = np.arange(0, (self.rows) * self.gap, self.gap)\n",
    "\n",
    "    def fill_matrix(self):\n",
    "\n",
    "        self.init_matrix()\n",
    "\n",
    "        for i in range(1, self.rows):                   #iterate through each cell of the matrix\n",
    "            for j in range(1, self.columns):\n",
    "\n",
    "                if self.seq1[i - 1] == self.seq2[j - 1]: #if the same, it's a match\n",
    "                    match_score = self.match\n",
    "                else:                                    #otherwise, penalize as a mismatch\n",
    "                    match_score = self.mismatch\n",
    "            \n",
    "                up = self.matrix[i - 1, j] + self.gap\n",
    "                left = self.matrix[i, j - 1] + self.gap\n",
    "                diagonal = self.matrix[i - 1, j - 1] + match_score\n",
    "\n",
    "                self.matrix[i, j] = max(diagonal, up, left)     #save the best score\n",
    "\n",
    "    def find_score(self):\n",
    "        #score is the bottom right cell of the matrix\n",
    "        return self.matrix[self.rows - 1 , self.columns - 1]\n",
    "    \n",
    "    def traceback(self):\n",
    "        i = self.rows  - 1\n",
    "        j = self.columns - 1\n",
    "\n",
    "        align1, align2 = \"\", \"\"\n",
    "        traceback_coords = []\n",
    "\n",
    "        while i > 0 and j > 0:\n",
    "            traceback_coords.append((i, j))\n",
    "            current = self.matrix[i, j]\n",
    "            x1 = self.seq1[i - 1]\n",
    "            x2 = self.seq2[j - 1]\n",
    "\n",
    "\n",
    "            if x1 == x2: #determine score\n",
    "                match_score = self.match\n",
    "            else:\n",
    "                match_score = self.mismatch\n",
    "\n",
    "            #diagonal\n",
    "            if current == self.matrix[i - 1, j - 1] + match_score:\n",
    "                align1 = x1 + align1\n",
    "                align2 = x2 + align2\n",
    "                i -= 1\n",
    "                j -= 1\n",
    "\n",
    "            #up\n",
    "            elif current == self.matrix[i - 1, j] + self.gap:\n",
    "                align1 = x1 + align1\n",
    "                align2 = \"-\" + align2\n",
    "                i -= 1\n",
    "\n",
    "            #left\n",
    "            else:\n",
    "                align1 = \"-\" + align1\n",
    "                align2 = x2 + align2\n",
    "                j -= 1\n",
    "\n",
    "        #fill rest with gaps if one side already aligned\n",
    "        while i > 0:\n",
    "            traceback_coords.append((i, j))\n",
    "            align1 = self.seq1[i - 1] + align1\n",
    "            align2 = \"-\" + align2\n",
    "            i -= 1\n",
    "\n",
    "        while j > 0:\n",
    "            traceback_coords.append((i, j))\n",
    "            align1 = \"-\" + align1\n",
    "            align2 = self.seq2[j - 1] + align2\n",
    "            j -= 1\n",
    "\n",
    "        self.traceback_coords = traceback_coords\n",
    "        return align1, align2\n",
    "\n",
    "    def display_matrix(self):\n",
    "        '''\n",
    "        Returns the scoring matrix with the traceback labeled in red\n",
    "        '''\n",
    "        df = pd.DataFrame(self.matrix) \n",
    "\n",
    "        styled_df = df.style.apply( lambda row: [\n",
    "            \"background-color: crimson\" \n",
    "            if (row.name, col_idx) in self.traceback_coords else \"\"\n",
    "            for col_idx in range(df.shape[1]) ], axis=1 )\n",
    "        \n",
    "\n",
    "        return styled_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70299ed8",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "example_sw = SmithWaterman(\"AGGTA\", \"ACGT\")\n",
    "print(f\"Smith Waterman Score: {example_sw.score}\\n\")\n",
    "print(f\"Smith Waterman Alignment:\\n{example_sw.alignment[0]}\\n{example_sw.alignment[1]}\")\n",
    "\n",
    "example_sw.display_matrix()\n",
    "\n",
    "example_nw = NeedlemanWunsch(\"AGGTA\", \"ACGT\")\n",
    "print(f\"Needleman-Wunsch Score: {example_nw.score}\\n\")\n",
    "print(f\"Needleman-Wunsch Alignment:\\n{example_nw.alignment[0]}\\n{example_nw.alignment[1]}\")\n",
    "\n",
    "example_nw.display_matrix()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3aba5e7",
   "metadata": {},
   "source": [
    "### Real world examples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68181bbe",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "human_hbb = read_fasta(\"data/HBB_HUMAN.fasta\")\n",
    "mouse_hbb = read_fasta(\"data/HBB1_MOUSE.fasta\")\n",
    "pigeon_hbb = read_fasta(\"data/HBB_COLLI.fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bcc84862",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "#Human Hemoglobin VS Mouse Hemoglobin\n",
    "\n",
    "humanVmouse_sm = SmithWaterman(human_hbb, mouse_hbb)\n",
    "print(f\"Smith Waterman Score: {humanVmouse_sm.score}\\n\")\n",
    "print(f\"Smith Waterman Alignment:\\n{humanVmouse_sm.alignment[0]}\\n{humanVmouse_sm.alignment[1]}\")\n",
    "\n",
    "humanVmouse_nw = NeedlemanWunsch(human_hbb, mouse_hbb)\n",
    "print(f\"Needleman-Wunsch Score: {humanVmouse_nw.score}\\n\")\n",
    "print(f\"Needleman-Wunsch Alignment:\\n{humanVmouse_nw.alignment[0]}\\n{humanVmouse_nw.alignment[1]}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aff4859c",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "#Human Hemoglobin VS Rock Dove Hemoglobin\n",
    "\n",
    "humanVpigeon_sw = SmithWaterman(human_hbb, pigeon_hbb)\n",
    "print(f\"Smith Waterman Score: {humanVpigeon_sw.score}\\n\")\n",
    "print(f\"Smith Waterman Alignment:\\n{humanVpigeon_sw.alignment[0]}\\n{humanVpigeon_sw.alignment[1]}\")\n",
    "\n",
    "humanVpigeon_nw = NeedlemanWunsch(human_hbb, pigeon_hbb)\n",
    "print(f\"Needleman-Wunsch Score: {humanVpigeon_nw.score}\\n\")\n",
    "print(f\"Needleman-Wunsch Alignment:\\n{humanVpigeon_nw.alignment[0]}\\n{humanVpigeon_nw.alignment[1]}\")"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
