{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e9e30dde",
   "metadata": {},
   "source": [
    "# Needleman-Wunsch Algorithm\n",
    "\n",
    "This algorithm, developed by Saul B. Needleman and Christian D. Wunsch, is utilized for global sequence alignment of amino acids and nucleotides."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9222d8f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8c2ad385",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_fasta(filename):\n",
    "    with open(filename, 'r') as f:\n",
    "        sequence = \"\"\n",
    "        for line in f:\n",
    "            line = line.strip()\n",
    "            if line.startswith(\">\"): #ignore first line with label\n",
    "                pass\n",
    "            else:\n",
    "                sequence += line\n",
    "    return sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "aa191f74",
   "metadata": {},
   "outputs": [],
   "source": [
    "class NeedlemanWunsch:\n",
    "    '''\n",
    "    Contains the Needleman-Wunsch score & matrix for two sequences.\n",
    "\n",
    "    Important Attributes:\n",
    "    score (int) : the score for the matching algorithm\n",
    "    alignment (tuple (str, str)) : the alignments for sequence 1 and 2\n",
    "\n",
    "    Important Methods:\n",
    "    display_matrix() : displays the scored matrix as a DataFrame\n",
    "                       where light blue is the traceback method.\n",
    "    '''\n",
    "    def __init__(self, seq1, seq2, match = 1, mismatch = -1, gap = -1):\n",
    "        self.seq1 = seq1\n",
    "        self.seq2 = seq2\n",
    "        self.match = match\n",
    "        self.mismatch = mismatch\n",
    "        self.gap = gap\n",
    "    \n",
    "        self.rows = len(seq1) + 1\n",
    "        self.columns = len(seq2) + 1\n",
    "\n",
    "        self.fill_matrix()\n",
    "\n",
    "        self.score = self.find_score()\n",
    "        self.alignment = self.traceback()\n",
    "\n",
    "    def init_matrix(self):\n",
    "        self.matrix = np.zeros((self.rows, self.columns), dtype = int)\n",
    "        self.matrix[0, :] = np.arange(0, (self.columns) * self.gap, self.gap)\n",
    "        self.matrix[:, 0] = np.arange(0, (self.rows) * self.gap, self.gap)\n",
    "\n",
    "    def fill_matrix(self):\n",
    "\n",
    "        self.init_matrix()\n",
    "\n",
    "        for i in range(1, self.rows):                   #iterate through each cell of the matrix\n",
    "            for j in range(1, self.columns):\n",
    "\n",
    "                if self.seq1[i - 1] == self.seq2[j - 1]: #if the same, it's a match\n",
    "                    match_score = self.match\n",
    "                else:                                    #otherwise, penalize as a mismatch\n",
    "                    match_score = self.mismatch\n",
    "            \n",
    "                up = self.matrix[i - 1, j] + self.gap\n",
    "                left = self.matrix[i, j - 1] + self.gap\n",
    "                diagonal = self.matrix[i - 1, j - 1] + match_score\n",
    "\n",
    "                self.matrix[i, j] = max(diagonal, up, left)     #save the best score\n",
    "\n",
    "    def find_score(self):\n",
    "        #score is the bottom right cell of the matrix\n",
    "        return self.matrix[self.rows - 1 , self.columns - 1]\n",
    "    \n",
    "    def traceback(self):\n",
    "        i = self.rows  - 1\n",
    "        j = self.columns - 1\n",
    "\n",
    "        align1, align2 = \"\", \"\"\n",
    "        traceback_coords = []\n",
    "\n",
    "        while i > 0 and j > 0:\n",
    "            traceback_coords.append((i, j))\n",
    "            current = self.matrix[i, j]\n",
    "            x1 = self.seq1[i - 1]\n",
    "            x2 = self.seq2[j - 1]\n",
    "\n",
    "\n",
    "            if x1 == x2: #determine score\n",
    "                match_score = self.match\n",
    "            else:\n",
    "                match_score = self.mismatch\n",
    "\n",
    "            #diagonal\n",
    "            if current == self.matrix[i - 1, j - 1] + match_score:\n",
    "                align1 = x1 + align1\n",
    "                align2 = x2 + align2\n",
    "                i -= 1\n",
    "                j -= 1\n",
    "\n",
    "            #up\n",
    "            elif current == self.matrix[i - 1, j] + self.gap:\n",
    "                align1 = x1 + align1\n",
    "                align2 = \"-\" + align2\n",
    "                i -= 1\n",
    "\n",
    "            #left\n",
    "            else:\n",
    "                align1 = \"-\" + align1\n",
    "                align2 = x2 + align2\n",
    "                j -= 1\n",
    "\n",
    "        #fill rest with gaps if one side already aligned\n",
    "        while i > 0:\n",
    "            traceback_coords.append((i, j))\n",
    "            align1 = self.seq1[i - 1] + align1\n",
    "            align2 = \"-\" + align2\n",
    "            i -= 1\n",
    "\n",
    "        while j > 0:\n",
    "            traceback_coords.append((i, j))\n",
    "            align1 = \"-\" + align1\n",
    "            align2 = self.seq2[j - 1] + align2\n",
    "            j -= 1\n",
    "\n",
    "        self.traceback_coords = traceback_coords\n",
    "        return align1, align2\n",
    "\n",
    "    def display_matrix(self):\n",
    "        '''\n",
    "        Returns the scoring matrix with the traceback labeled in red\n",
    "        '''\n",
    "        df = pd.DataFrame(self.matrix) \n",
    "\n",
    "        styled_df = df.style.apply( lambda row: [\n",
    "            \"background-color: crimson\" \n",
    "            if (row.name, col_idx) in self.traceback_coords else \"\"\n",
    "            for col_idx in range(df.shape[1]) ], axis=1 )\n",
    "        \n",
    "\n",
    "        return styled_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f824b53b",
   "metadata": {},
   "source": [
    "Short Sequence Example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "67113183",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score: 1\n",
      "\n",
      "Alignment:\n",
      "AGGTA\n",
      "ACGT-\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_aab16_row1_col1, #T_aab16_row2_col2, #T_aab16_row3_col3, #T_aab16_row4_col4, #T_aab16_row5_col4 {\n",
       "  background-color: crimson;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_aab16\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th id=\"T_aab16_level0_col0\" class=\"col_heading level0 col0\" >0</th>\n",
       "      <th id=\"T_aab16_level0_col1\" class=\"col_heading level0 col1\" >1</th>\n",
       "      <th id=\"T_aab16_level0_col2\" class=\"col_heading level0 col2\" >2</th>\n",
       "      <th id=\"T_aab16_level0_col3\" class=\"col_heading level0 col3\" >3</th>\n",
       "      <th id=\"T_aab16_level0_col4\" class=\"col_heading level0 col4\" >4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_aab16_level0_row0\" class=\"row_heading level0 row0\" >0</th>\n",
       "      <td id=\"T_aab16_row0_col0\" class=\"data row0 col0\" >0</td>\n",
       "      <td id=\"T_aab16_row0_col1\" class=\"data row0 col1\" >-1</td>\n",
       "      <td id=\"T_aab16_row0_col2\" class=\"data row0 col2\" >-2</td>\n",
       "      <td id=\"T_aab16_row0_col3\" class=\"data row0 col3\" >-3</td>\n",
       "      <td id=\"T_aab16_row0_col4\" class=\"data row0 col4\" >-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_aab16_level0_row1\" class=\"row_heading level0 row1\" >1</th>\n",
       "      <td id=\"T_aab16_row1_col0\" class=\"data row1 col0\" >-1</td>\n",
       "      <td id=\"T_aab16_row1_col1\" class=\"data row1 col1\" >1</td>\n",
       "      <td id=\"T_aab16_row1_col2\" class=\"data row1 col2\" >0</td>\n",
       "      <td id=\"T_aab16_row1_col3\" class=\"data row1 col3\" >-1</td>\n",
       "      <td id=\"T_aab16_row1_col4\" class=\"data row1 col4\" >-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_aab16_level0_row2\" class=\"row_heading level0 row2\" >2</th>\n",
       "      <td id=\"T_aab16_row2_col0\" class=\"data row2 col0\" >-2</td>\n",
       "      <td id=\"T_aab16_row2_col1\" class=\"data row2 col1\" >0</td>\n",
       "      <td id=\"T_aab16_row2_col2\" class=\"data row2 col2\" >0</td>\n",
       "      <td id=\"T_aab16_row2_col3\" class=\"data row2 col3\" >1</td>\n",
       "      <td id=\"T_aab16_row2_col4\" class=\"data row2 col4\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_aab16_level0_row3\" class=\"row_heading level0 row3\" >3</th>\n",
       "      <td id=\"T_aab16_row3_col0\" class=\"data row3 col0\" >-3</td>\n",
       "      <td id=\"T_aab16_row3_col1\" class=\"data row3 col1\" >-1</td>\n",
       "      <td id=\"T_aab16_row3_col2\" class=\"data row3 col2\" >-1</td>\n",
       "      <td id=\"T_aab16_row3_col3\" class=\"data row3 col3\" >1</td>\n",
       "      <td id=\"T_aab16_row3_col4\" class=\"data row3 col4\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_aab16_level0_row4\" class=\"row_heading level0 row4\" >4</th>\n",
       "      <td id=\"T_aab16_row4_col0\" class=\"data row4 col0\" >-4</td>\n",
       "      <td id=\"T_aab16_row4_col1\" class=\"data row4 col1\" >-2</td>\n",
       "      <td id=\"T_aab16_row4_col2\" class=\"data row4 col2\" >-2</td>\n",
       "      <td id=\"T_aab16_row4_col3\" class=\"data row4 col3\" >0</td>\n",
       "      <td id=\"T_aab16_row4_col4\" class=\"data row4 col4\" >2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_aab16_level0_row5\" class=\"row_heading level0 row5\" >5</th>\n",
       "      <td id=\"T_aab16_row5_col0\" class=\"data row5 col0\" >-5</td>\n",
       "      <td id=\"T_aab16_row5_col1\" class=\"data row5 col1\" >-3</td>\n",
       "      <td id=\"T_aab16_row5_col2\" class=\"data row5 col2\" >-3</td>\n",
       "      <td id=\"T_aab16_row5_col3\" class=\"data row5 col3\" >-1</td>\n",
       "      <td id=\"T_aab16_row5_col4\" class=\"data row5 col4\" >1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x7ca650fce990>"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "example = NeedlemanWunsch(\"AGGTA\", \"ACGT\")\n",
    "print(f\"Score: {example.score}\\n\")\n",
    "print(f\"Alignment:\\n{example.alignment[0]}\\n{example.alignment[1]}\")\n",
    "\n",
    "example.display_matrix()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec6b6998",
   "metadata": {},
   "source": [
    "## Real World Examples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "b84844c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "human_hbb = read_fasta(\"data/HBB_HUMAN.fasta\")\n",
    "mouse_hbb = read_fasta(\"data/HBB1_MOUSE.fasta\")\n",
    "pigeon_hbb = read_fasta(\"data/HBB_COLLI.fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51d9df06",
   "metadata": {},
   "source": [
    "Human Hemoglobin VS Mouse Hemoglobin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "cadc4a4f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score: 89\n",
      "\n",
      "Alignment:\n",
      "MVHLTPEEKSAVTALWGKVNVDEVGGEALGRLLVVYPWTQRFFESFGDLSTPDAVMGNPKVKAHGKKVLGAFSDGLAHLDNLKGTFATLSELHCDKLHVDPENFRLLGNVLVCVLAHHFGKEFTPPVQAAYQKVVAGVANALAHKYH\n",
      "MVHLTDAEKAAVSCLWGKVNSDEVGGEALGRLLVVYPWTQRYFDSFGDLSSASAIMGNAKVKAHGKKVITAFNDGLNHLDSLKGTFASLSELHCDKLHVDPENFRLLGNMIVIVLGHHLGKDFTPAAQAAFQKVVAGVATALAHKYH\n"
     ]
    }
   ],
   "source": [
    "humanVmouse = NeedlemanWunsch(human_hbb, mouse_hbb)\n",
    "print(f\"Score: {humanVmouse.score}\\n\")\n",
    "print(f\"Alignment:\\n{humanVmouse.alignment[0]}\\n{humanVmouse.alignment[1]}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3c5d899",
   "metadata": {},
   "source": [
    "Human Hemoglobin VS Rock Dove Hemoglobin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "dcc3fb53",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score: 56\n",
      "\n",
      "Alignment:\n",
      "MVHLTPEEKSAVTALWGKVNV-DEVGGEALGRLLVVYPWTQRFFESFGDLSTPDAVMGNPKVKAHGKKVLGAFSDGLAHLDNLKGTFATLSELHCDKLHVDPENFRLLGNVLVCVLAHHFGKEFTPPVQAAYQKVVAGVANALAHKYH\n",
      "-VHWSAEEKQLITSIWGKVNVAD-CGAEALARLLIVYPWTQRFFSSFGNLSSATAISGNPNVKAHGKKVLTSFGDAVKNLDNIKGTFAQLSELHCDKLHVDPENFRLLGDILVIILAAHFGKDFTPECQAAWQKLVRVVAHALARKYH\n"
     ]
    }
   ],
   "source": [
    "humanVpigeon = NeedlemanWunsch(human_hbb, pigeon_hbb)\n",
    "print(f\"Score: {humanVpigeon.score}\\n\")\n",
    "print(f\"Alignment:\\n{humanVpigeon.alignment[0]}\\n{humanVpigeon.alignment[1]}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "msse-python",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
